### PR TITLE
[xcvrd] Modify to support regular expression when parsing the key in media_settings.json

### PR DESCRIPTION
**Pull request by DennisChiuEC**
#### Description

Let media_settings.json can use regular expression to match 'vendor key' or 'media key'
when define a set of transceiver to use the same SI value.

#### Motivation and Context

Make the definition of SI value for a set of vendor or media can be more flexible

#### How Has This Been Tested?

`root@sonic:~# cat /var/log/swss/swss.rec | grep preemphasis`
`2024-04-17.01:53:10.594465|PORT_TABLE:Ethernet0|SET|preemphasis:0x0c8418,0x0c8418,0x0c8418,0x0c8418`
`2024-04-17.01:53:11.827324|PORT_TABLE:Ethernet4|SET|preemphasis:0x0c8418,0x0c8418,0x0c8418,0x0c8418`

Example of media_settings.json
`    "GLOBAL_MEDIA_SETTINGS": {`
`        "1-32": {`
`            "VENDOR.*-PN.*": {`
`                "preemphasis": {`
`                    "lane0": "0x0c8418",`
`                    "lane1": "0x0c8418",`
`                    "lane2": "0x0c8418",`
`                    "lane3": "0x0c8418",`
`                    "lane4": "0x0c8418"`
`                }`
`            }`
`        }`

### CODE REVIEW

**Comment by prgeor** [comment]
@DennisChiuEC Please add unit test

**Review by prgeor — COMMENTED** [review]
@moshemos @tshalvi Please review

**Comment by DennisChiuEC** [comment]
OK, I will add the unit test case.

**Comment by keboliu** [comment]
@tshalvi would you please review?

**Review by tshalvi — APPROVED** [review]

File: sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py [hunk 1]
```diff
@@ -193,31 +194,19 @@ def get_media_settings_value(physical_port, key):
 
             # If there is a match in the global profile for a media type,
             # fetch those values
-            if key[VENDOR_KEY] in media_dict: # e.g: 'AMPHENOL-1234'
-                if is_si_per_speed_supported(media_dict[key[VENDOR_KEY]]):
-                    if key[LANE_SPEED_KEY] is not None and key[LANE_SPEED_KEY] in media_dict[key[VENDOR_KEY]]: # e.g: 'speed:400GAUI-8'
-                        return media_dict[key[VENDOR_KEY]][key[LANE_SPEED_KEY]]
+            for dict_key in media_dict.keys():
```
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py:197]: @DennisChiuEC can you write a function `get_media_settings`

File: sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py [hunk 2]
```diff
@@ -235,31 +224,19 @@ def get_media_settings_value(physical_port, key):
...
-                    return media_dict[key[VENDOR_KEY].split('-')[0]][key[LANE_SPEED_KEY]]
-                else:
-                    return {}
-            else:
-                return media_dict[key[VENDOR_KEY].split('-')[0]]
-        elif key[MEDIA_KEY] in media_dict:
-            if is_si_per_speed_supported(media_dict[key[MEDIA_KEY]]):
-                if key[LANE_SPEED_KEY] is not None and key[LANE_SPEED_KEY] in media_dict[key[MEDIA_KEY]]:
-                    return media_dict[key[MEDIA_KEY]][key[LANE_SPEED_KEY]]
+        for dict_key in media_dict.keys():
```
Comment (prgeor) [inline sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py:227]: @DennisChiuEC can you write a function `get_media_settings`

Comment (DennisChiuEC) [reply sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py]: Done

File: sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py [hunk 3]
```diff
@@ -168,6 +169,19 @@ def get_media_settings_value(physical_port, key):
     media_dict = {}
     default_dict = {}
 
+    def get_media_settings(key, media_dict):
+        for dict_key in media_dict.keys():
+            if ( re.match(dict_key, key[VENDOR_KEY]) or re.match(dict_key, key[VENDOR_KEY].split('-')[0]) # e.g: 'AMPHENOL-1234'
```
Comment (mihirpat1) [inline sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py:174]:
```suggestion
            if (re.match(dict_key, key[VENDOR_KEY]) or \
                re.match(dict_key, key[VENDOR_KEY].split('-')[0]) # e.g: 'AMPHENOL-1234'
```

File: sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py [hunk 4]
```diff
@@ -168,6 +169,19 @@ def get_media_settings_value(physical_port, key):
...
+                 or re.match(dict_key, key[MEDIA_KEY]) ): # e.g: 'QSFP28-40GBASE-CR4-1M'
+                if is_si_per_speed_supported(media_dict[dict_key]):
+                    if key[LANE_SPEED_KEY] is not None and key[LANE_SPEED_KEY] in media_dict[dict_key]: # e.g: 'speed:400GAUI-8'
+                        return media_dict[dict_key][key[LANE_SPEED_KEY]]
+                    else:
+                        return {}
```
Comment (mihirpat1) [inline sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py:180]: Why are we returning `{}` and not None?

Comment (DennisChiuEC) [reply sonic-xcvrd/xcvrd/xcvrd_utilities/media_settings_parser.py:181]: It was added by the previous PR.
I just defined a function to simplify the code.

**Review by mihirpat1 — APPROVED** [review]

**Review by prgeor — COMMENTED** [review]
@tshalvi can you please review?

**Review by prgeor — APPROVED** [review]

**Comment by longhuan-cisco** [comment]
@prgeor could we add `Request for 202405 Branch` to cherry-pick this PR?

**Comment by mihirpat1** [comment]
@bingwang-ms Can you please help to cherry-pick this to 202405?

**Comment by longhuan-cisco** [comment]
@prgeor 

Could we also cherry-pick this PR to 202311?

PR https://github.com/sonic-net/sonic-platform-daemons/pull/533 (which was requested for 202311 cherry-pick) depends on the regular expression support introduced by this PR.

**Comment by mihirpat1** [comment]
> @prgeor
> 
> Could we also cherry-pick this PR to 202311?
> 
> PR #533 (which was requested for 202311 cherry-pick) depends on the regular expression support introduced by this PR.

@longhuan-cisco Can you please help to create a separate 202311 PR for PR#533 since the dependent PR#471 is now cherry-picked to 202311.

**Comment by longhuan-cisco** [comment]
> @longhuan-cisco Can you please help to create a separate 202311 PR for PR#533 since the dependent PR#471 is now cherry-picked to 202311.

raised https://github.com/sonic-net/sonic-platform-daemons/pull/545 for it

**Pull request merged**